packages/connect-migrate/src/cli.ts

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* eslint-disable no-console -- We use this to output results. */
+import { replacePackageJSONReferences } from "./replacement-map";
+import fastGlob from "fast-glob";
+import { readFile, writeFile } from "node:fs/promises";
+import { existsSync } from "node:fs";
+import * as path from "node:path";
+import { createRequire } from "node:module";
+import * as url from "node:url";
+import assert from "node:assert";
+import process from "node:process";
+import { spawnSync } from "node:child_process";
+
+const __dirname = url.fileURLToPath(new URL(".", import.meta.url));
+const require = createRequire(import.meta.url);
+const jscodeshiftExecutable = require.resolve(".bin/jscodeshift");
+
+// We reference src here so jscodeshift can digest the transform
+// as it is, since jscodeshift can't seem to handle compiled ts.
+const transformerDirectory = path.join(__dirname, "../", "src", "transforms");
+const transformerPath = path.join(transformerDirectory, `modify-imports.ts`);
+
+async function replacePackageReferences() {
+  const dir = process.cwd();
+  const packageExists = existsSync(path.join(dir, "package.json"));
+
+  function executeInContext(command: string, args: string[]) {
+    spawnSync(command, args, {
+      cwd: dir,
+      shell: true,
+      stdio: ["pipe", "inherit", "inherit"],
+    });
+  }
+
+  assert(
+    packageExists,
+    "Could not find package.json. This command must be run in a directory with a package.json file.",
+  );
+  let packageManager: "pnpm" | "npm" | "yarn" | undefined = undefined;
+  if (existsSync(path.join(dir, "pnpm-lock.yaml"))) {
+    packageManager = "pnpm";
+  } else if (existsSync(path.join(dir, "package-lock.json"))) {
+    packageManager = "npm";
+  } else if (existsSync(path.join(dir, "yarn.lock"))) {
+    packageManager = "yarn";
+  }
+  assert(
+    packageManager !== undefined,
+    "Could not determine package manager based on lock file. Make sure you have a lock file (yarn.lock, package-lock.json, or pnpm-lock.yaml) in your project.",
+  );
```

Comment:
We cannot use assertions for flow control. It's easy to run the command in the wrong directory, and we have to consider it to be part of the regular flow, not an exceptional case.

In general, command line tools have to:
- exit with status 0 on success
- only crash with stack trace on an exceptional error
- print a short error message (no stack traces) and usage information on stderr and exit with non-zero code for invalid arguments
- print a short error message and exit with non-zero code in case of an error
- print usage information for the flag `-h` or `--help`
- print the version for the flag `--version` (or `-v` unless taken for verbose output)

Here is an example for parsing a command line:

```ts
interface CommandLineArgs {
  ok: true;
  help: boolean;
  version: boolean;
}

interface CommandLineError {
  ok: false;
  errorMessage: string;
}

export function parseCommandLineArgs(args: string[]): CommandLineArgs | CommandLineError {
  const parsed = {
    version: false,
    help: false
  };
  while (args.length > 0) {
    const arg = args.shift();
    switch (arg) {
      case "-h":
      case "--help":
        if (parsed.help) {
          break;
        }
        parsed.help = true;
        break;
      case "--version":
        parsed.version = true;
        break;
      default:
        return {
          ok: false,
          errorMessage: `unknown argument ${arg}`
        };
    }
  }
  return {...parsed, ok: true};
}
```

Used as:

```ts
import { parseCommandLineArgs } from "./arguments";
import process from "node:process";

const args = parseCommandLineArgs(process.argv.slice(2));
if (!args.ok) {
  process.stderr.write(`${args.errorMessage}\n`);
  process.exit(1);
}
if (args.version) {
  // todo: use version from package.json
  process.stdout.write("connect-migrate v0.1.0\n");
  process.exit(0);
}
if (args.help) {
  process.stdout.write(`USAGE: connect-migrate\n`);
  process.stdout.write(`Updates references to connect-es packages in your project to use @connectrpc.\n`);
  process.exit(0);
}

// todo: logic
process.exit(0);
```

This scales up to everything we'll need here, and makes testing trivial.